packages/projection/src/sinks/inMemory/InMemoryStabilityWindowBuffer.ts

```diff
@@ -1,39 +1,45 @@
-import { BehaviorSubject, EMPTY, Observable } from 'rxjs';
-import { Cardano } from '@cardano-sdk/core';
-import { RollForwardEvent } from '../../types';
+import { BehaviorSubject, tap } from 'rxjs';
+import { Cardano, ChainSyncEventType } from '@cardano-sdk/core';
 import { StabilityWindowBuffer } from '../types';
+import { UnifiedProjectorOperator } from '../../types';
 import { WithNetworkInfo } from '../../operators';
 
 export class InMemoryStabilityWindowBuffer<E extends WithNetworkInfo> implements StabilityWindowBuffer<E> {
   readonly #blocks: Cardano.Block[] = [];
   readonly tip$ = new BehaviorSubject<Cardano.Block | 'origin'>('origin');
   readonly tail$ = new BehaviorSubject<Cardano.Block | 'origin'>('origin');
+  readonly handleEvents: UnifiedProjectorOperator<E, E>;
 
-  deleteBlock(block: Cardano.Block): Observable<void> {
-    for (let i = this.#blocks.length - 1; i >= 0; i--) {
-      const bufferBlock = this.#blocks[i];
-      if (bufferBlock.header.hash === block.header.hash) {
-        this.#blocks.splice(i, 1);
-        break;
-      }
-    }
-    this.tip$.next(this.#blocks[this.#blocks.length - 1] || 'origin');
-    const newTail = this.#blocks[0] || 'origin';
-    this.#setTail(newTail);
-    return EMPTY;
+  constructor() {
+    this.handleEvents = (evt$) =>
+      evt$.pipe(
+        tap(({ eventType, block, genesisParameters: { securityParameter } }) => {
+          if (eventType === ChainSyncEventType.RollForward) {
+            // clear blocks that are past stability window
+            while (this.#blocks.length > securityParameter) this.#blocks.shift();
+            // add current block to cache and return the event unchanged
+            this.#blocks.push(block);
+            this.tip$.next(block);
+            this.#setTail(this.#blocks[0]);
+          } else if (eventType === ChainSyncEventType.RollBackward) {
+            for (let i = this.#blocks.length - 1; i >= 0; i--) {
```

Comment:
I will start with the assumption that a rollback always happens on the tip$. If that is correct, then iterating the blocks backwards, trying to find the rollback block does not make sense: It's either the tip$ or nothing.
On the other hand, if the rollback event could be on an older block, all blocks newer than it should be deleted.
The for loop looks for the rollback block and removes it, leaving newer blocks untouched.

Reply:
In context of stability window buffer it will always pop the tip - a rollback can happen to an older block, but this gets called for each block.

So yeah, this implementation isn't great, but in-memory buffer is out of scope of this PR, this was just a refactor. We'll probably only be using this buffer in tests, but feel free to open a ticket to fix this.

Reply:
90595eac41c9eececcd58fe2be21f91b95bf8f80